subprojects/core/src/main/java/org/gradle/internal/build/BuildOperationFiringBuildWorkPreparer.java

```diff
@@ -120,104 +132,87 @@ void populateTaskGraph() {
-        if (nodes.isEmpty()) {
-            return Collections.emptyList();
+        private List<? extends NodeIdentity> findDependencies(Node node) {
+            return findDependencies(node.getDependencySuccessors(), Node::getDependencySuccessors);
         }
-        return FluentIterable.from(nodes)
-            .filter(seen::add)
-            .transformAndConcat(
-                node -> node instanceof TaskNode
-                    ? ImmutableSet.of((TaskNode) node)
-                    : traverseNonTasks(requireNonNull(traverser.apply(node)), traverser, seen)
-            );
-    }
 
-    private static List<CalculateTaskGraphBuildOperationType.TaskIdentity> taskIdentifiesOf(Collection<Node> nodes) {
-        if (nodes.isEmpty()) {
-            return Collections.emptyList();
-        }
-        return FluentIterable.from(nodes)
-            .filter(TaskNode.class)
-            .transform(BuildOperationFiringBuildWorkPreparer::toIdentity)
-            .toList();
-    }
+        private List<? extends NodeIdentity> findDependencies(Collection<Node> nodes, Function<? super Node, ? extends Collection<Node>> traverser) {
+            if (nodes.isEmpty()) {
+                return Collections.emptyList();
+            }
 
-    private static CalculateTaskGraphBuildOperationType.TaskIdentity toIdentity(TaskNode n) {
-        return new PlannedTaskIdentity(n.getTask().getTaskIdentity());
-    }
+            return findIdentifiedNodes(nodes, traverser, newSetFromMap(new IdentityHashMap<>()))
+                .collect(Collectors.toList());
+        }
 
-    private static List<String> toTaskPaths(Set<Task> tasks) {
-        return ImmutableSortedSet.copyOf(Collections2.transform(tasks, Task::getPath)).asList();
-    }
+        private Stream<NodeIdentity> findIdentifiedNodes(Collection<Node> nodes, Function<? super Node, ? extends Collection<Node>> traverser, Set<Node> seen) {
+            if (nodes.isEmpty()) {
+                return Stream.empty();
+            }
 
-    private static class PlannedTaskIdentity implements CalculateTaskGraphBuildOperationType.TaskIdentity {
+            return nodes.stream()
+                .filter(seen::add)
+                .flatMap(node -> {
+                    ToPlannedNodeConverter converter = getConverter(node);
+                    // We do not check for the same execution plan, because a node can depend on nodes from other plans
+                    return converter != null
+                        ? Stream.of(converter.getNodeIdentity(node))
+                        : findIdentifiedNodes(traverser.apply(node), traverser, seen);
+                });
+        }
 
-        private final TaskIdentity<?> delegate;
+        private ToPlannedNodeConverter getConverter(Node node) {
+            Class<? extends Node> nodeType = node.getClass();
+            ToPlannedNodeConverter converter = convertersByNodeType.get(nodeType);
+            if (converter != null) {
```

Comment:
That wouldn't probably be noticeable now, but an unsupported node type would force the traversal of all converters. This can become more expensive as we add new converters.